drivers/usb/usbport/device.c

```diff
@@ -885,6 +885,76 @@ USBPORT_AbortTransfers(IN PDEVICE_OBJECT FdoDevice,
     }
 }
 
+PUSB2_TT_EXTENSION
+NTAPI
+USBPORT_GetTt(IN PDEVICE_OBJECT FdoDevice,
+              IN PUSBPORT_DEVICE_HANDLE HubDeviceHandle,
+              OUT PUSHORT OutPort,
+              OUT PUSBPORT_DEVICE_HANDLE * OutHubDeviceHandle)
+{
+    PUSBPORT_DEVICE_HANDLE DeviceHandle = HubDeviceHandle;
+    ULONG TtCount;
+    PLIST_ENTRY Entry;
+    PUSB2_TT_EXTENSION TtExtension = NULL;
+
+    DPRINT("USBPORT_GetTt: HubDeviceHandle - %p\n", HubDeviceHandle);
+
+    *OutHubDeviceHandle = NULL;
+
+    while (DeviceHandle->DeviceSpeed != UsbHighSpeed)
+    {
+        DPRINT("USBPORT_GetTt: DeviceHandle - %p, DeviceHandle->PortNumber - %X\n",
+               DeviceHandle,
+               DeviceHandle->PortNumber);
+
+        *OutPort = DeviceHandle->PortNumber;
+
+        DeviceHandle = DeviceHandle->HubDeviceHandle;
+
+        if (!DeviceHandle)
+            return NULL;
+    }
+
+    TtCount = DeviceHandle->TtCount;
+
+    if (!TtCount)
+        return NULL;
+
+    Entry = DeviceHandle->TtList.Flink;
+
+    if (Entry != &DeviceHandle->TtList)
```

Comment:
I'd like to write this as something like

```
while (Entry != &DeviceHandle->TtList)
{
    TtExtension = CONTAINING_RECORD(Entry, USB2_TT_EXTENSION, Link);
    Entry = Entry->Flink;
    if (TtExtension->TtNumber == *OutPort)
        break;
    TtExtension = NULL;
}
if (TtExtension != NULL)
    *OutHubDeviceHandle = DeviceHandle;
```

But I'm worried about the slightly different logic resulting from that.
Any suggestions for improvements that don't change the logic?

Reply:
Supposing that DeviceHandle->TtCount has been consistently increased when stuff has been added to this list (and decreased when stuff has been removed from it), *THEN* your code behaves identically to the original one.
Also, in the case where there is only one element in this list, the situation whether or not we break out of the loop because TtExtension->TtNumber == *OutPort is TRUE, or is not, should not matter: because in the case the test is FALSE, we just break out of the loop afterwards because Entry now points to the TtList head.

Reply:
The one element case right now will return the first TtExtension no matter its TtNumber. With my change it would return it only if the TtNumber matches.
Note the difference between breaking out of the loop with TtExtension=NULL vs TtExtension set.